lib/puppet/application/catalog.rb

```diff
@@ -1,4 +1,7 @@
 require 'puppet/application/indirection_base'
 
 class Puppet::Application::Catalog < Puppet::Application::IndirectionBase
+  option("--run_mode " + _("MODE")) do |arg|
+    self.class.run_mode arg.intern
+  end
```

Comment:
Ah so `--run_mode` is already a global command line option: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/settings.rb#L291-L295, so this can be deleted. We don't document it in general, because changing `run_mode` dynamically is fragile for puppet in general. See https://groups.google.com/d/topic/puppet-dev/_ebwwlW6S9s/discussion. For this specific case I think it's ok to document. That can be done by replacing the above with:

```ruby
  face = Puppet::Face[:catalog, :current]
  face.option("--run_mode " + _("MODE")) do
    summary _("To compile catalogs as a master use 'puppet catalog find --run_mode master'")
  end
```
which will result in:

```
$ bx puppet catalog --help
...
  --run_mode MODE                - To compile catalogs as a master use 'puppet
                                   catalog find --run_mode master'
```

Reply:
I had found that `--run_mode` global but it doesn't seem to apply here:

```sh
$ bundle exec puppet catalog find --debug --run_mode master mynode.example.com | grep run_mode
Debug: Runtime environment: puppet_version=6.5.0, ruby_version=2.5.0, run_mode=master, default_encoding=UTF-8

$ git checkout master
Switched to branch 'master'

$ bundle exec puppet catalog find --debug --run_mode master mynode.example.com | grep run_mode
Error: Could not parse application options: invalid option: --run_mode
```

Following back up the stack, it appears the only place the global option parsing could be brought in is [`Puppet::Util::CommandLine#execute`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/command_line.rb#L72), but clearly that's not working as expected.

After some digging, I think I've found the cause:

Putting `git bisect` to the test, bisecting f4e229ee (when `--run_mode` was introduced) and `master`, turns out `--run_mode` was broken years ago with commit 7a8577d.

The difference between the good and bad output:

```sh
$ bundle exec puppet config print all --run_mode master | grep confdir
Error: Could not parse application options: invalid option: --run_mode

$ git checkout HEAD^
Previous HEAD position was 7a8577dd10... (Maint) Separate discovery of subcommand name from args
HEAD is now at 46f6725402... (Maint) Simplify logic to find subcommand name

$ bundle exec puppet config print all --run_mode master | grep confdir
confdir = /Users/alansmith/.puppet
```

It looks like 7a8577d switched from a stored `args` variable to a "getter" method which computed the `args` array. This is significant because `Puppet::Settings#parse_global_options` was modifying the variable to remove `--run_mode` and prevent future option parsing from choking on it. I presume this was never noticed because it's an undocumented option and doesn't have unit tests to validate its successful parsing.

Given all of this, I'm not sure how to proceed. It would seem the code for the global `--run_mode` is dead and should either be restored to its former glory -- in which case the code in this PR can go away -- or ripped out.

Reply:
Nice debugging! That's kind of amazing how long it's been broken. I think our options are mostly:

1. Reinstate the global `--run_mode <mode>` option
2. Modify `puppet catalog` to always run in the `:master` run mode, like `puppet lookup` does, or make it conditional based on the `find` action and/or `terminus`.
3. Create a new action, eg `:compile`, that that always sets `:master` run mode and delegates to `:find`

I'm not a huge fan of (1) because `run_mode` is terrible, and if it's been dead for 6 years, better to leave it that way.

(2) is likely to break anyone using `puppet catalog find --terminus rest`, as changing the `run_mode` can cause settings to be resolved differently, eg if `server` is defined in both `[main]` and `[server]`.

I think (3) is probably the best option. Something like this should work:

```
  action(:compile) do
    summary _("Compile a catalog.")
    arguments _("[<key>]")
    when_invoked do |*args|
      Puppet.settings.preferred_run_mode = :master
      Puppet::Face[:catalog, :current].find(*args)
    end
  end
```

Reply:
I think you're right, (3) is the best. But there's a slight issue with that:

`Puppet::Application` makes calls to lookup the preferred/current `run_mode` before the face is eval'd. This results in the `run_mode=user` you see below, followed by a stacktrace added after `when_invoked` in your `:compile` action:

```sh
$ bundle exec puppet catalog compile --debug mynode | head
Debug: Runtime environment: puppet_version=6.4.0, ruby_version=2.3.3, run_mode=user, default_encoding=UTF-8
Debug: /Users/alansmith/code/puppet/lib/puppet/interface/action.rb+eval[wrapper]:261:in `compile'
lib/puppet/application/face_base.rb:247:in `main'
lib/puppet/application.rb:390:in `run_command'
lib/puppet/application.rb:382:in `block in run'
lib/puppet/util.rb:671:in `exit_on_fail'
lib/puppet/application.rb:382:in `run'
lib/puppet/util/command_line.rb:139:in `run'
lib/puppet/util/command_line.rb:77:in `execute'
bin/puppet:5:in `<top (required)>'
```

In short, inside an `action` block is too late to change the `run_mode` for it to be seen globally. I mention this in my commit message on this PR. I couldn't find a way around this without a hack or refactor.

Reply:
Changing `run_mode` at runtime is generally not great, which is why most of the time is set for the entire application via `run_mode :master`. But for this issue it should work ok. I copied the above `compile` action:


```diff
diff --git a/lib/puppet/face/catalog.rb b/lib/puppet/face/catalog.rb
index c04c0a6eb1..5b0575876e 100644
--- a/lib/puppet/face/catalog.rb
+++ b/lib/puppet/face/catalog.rb
@@ -31,6 +31,21 @@ Puppet::Indirector::Face.define(:catalog, '0.0.1') do
     Puppet::Resource::Catalog object.
   EOT
 
+  action(:compile) do
+    summary _("Compile a catalog.")
+    description <<-'EOT'
+     ...
+    EOT
+    arguments _("[<key>]")
+    returns <<-'EOT'
+      ..
+    EOT
+    when_invoked do |*args|
+      Puppet.settings.preferred_run_mode = :master
+      Puppet::Face[:catalog, :current].find(*args)
+    end
+  end
+
   action(:apply) do
     summary "Find and apply a catalog."
     description <<-'EOT'
```

And it seems to work ok for me (both ruby 2.5.3 and 2.3.8):

```
$ bundle exec puppet catalog compile --debug mynode --codedir ~/work/pupperware/volumes/code
Debug: Runtime environment: puppet_version=6.5.0, ruby_version=2.5.3, run_mode=user, default_encoding=UTF-8
Debug: Loading external facts from /Users/josh/.puppetlabs/opt/puppet/cache/facts.d
...
Debug: Reset text domain to :production
Debug: importing '/Users/josh/work/pupperware/volumes/code/environments/production/modules/mymodule/manifests/init.pp' in environment production
Debug: Automatically imported mymodule from mymodule into production
Debug: Could not find library 'shadow' required to enable feature 'libshadow'
Debug: Could not find library 'selinux' required to enable feature 'selinux'
Notice: Compiled catalog for mynode in environment production in 0.41 seconds
{
  "tags": [
    "settings",
    "default",
    "mymodule",
    "node",
    "class"
...
```

Reply:
Also it does look like `run_mode` is set correctly by the time the compile happens. I added a `notify { "$settings::pidfile": }` resource, and verified it references `master` run mode:

```
    {
      "type": "Notify",
      "title": "/Users/josh/.puppetlabs/var/run/master.pid",
      ...
    }
```

Reply:
You’re right all appears to be fine, but I’m afraid there’s some edge case we’re not thinking of that will behave unexpectedly when the ‘run_mode’ is changed part way through the execution. And by “part way” I mean after ‘Puppet::Application’ has saved one value and ‘Puppet::Face’ comes along later and changes the global setting — the global setting that gets cached in some places.

Given that we’re shoehorning in this functionality, I want to be careful that it’s not going to cause more headache down the road.

Reply:
Yeah, we're unfortunately already in that boat for any application that's sets the `preferred_run_mode` while running. Ideally `run_mode` would be immutable and set before settings are initialized, but we're not there yet.

Reply:
@cloakedcode did you want to update this PR or should I?

Reply:
I'll do that right now!